waku/waku_filter_v2/subscriptions.nim

```diff
@@ -6,48 +6,161 @@ else:
+              maxPeers: maxFilterPeers,
+              maxCriteriaPerPeer: maxFilterCriteriaPerPeer
+            )
+
+proc isSubscribed*(s: var FilterSubscriptions, peerId: PeerID): bool =
+  s.peersSubscribed.withValue(peerId, data):
+    return Moment.now() - data.lastSeen <= s.subscriptionTimeout
+
+  return false
+
+proc subscribedPeerCount*(s: FilterSubscriptions): uint =
+  return cast[uint](s.peersSubscribed.len())
 
-  var subscribedPeers: seq[PeerID]
+proc getPeerSubscriptions*(s: var FilterSubscriptions, peerId: PeerID): seq[FilterCriterion] =
+  ## Get all pubsub-content topics a peer is subscribed to
+  var subscribedContentTopics: seq[FilterCriterion] = @[]
+  s.peersSubscribed.withValue(peerId, data):
+    if data.criteriaCount == 0:
+      return subscribedContentTopics
 
-  # TODO: for large maps, this can be optimized using a reverse index
-  for (peerId, criteria) in subscriptions.pairs():
-    if filterCriterion in criteria:
-      subscribedPeers.add(peerId)
+    for filterCriterion, subscribedPeers in s.subscriptions.mpairs:
+      if peerId in subscribedPeers:
+        subscribedContentTopics.add(filterCriterion)
 
-  subscribedPeers
+  return subscribedContentTopics
 
-proc removePeer*(subscriptions: var FilterSubscriptions, peerId: PeerID) =
+proc findSubscribedPeers*(s: var FilterSubscriptions, pubsubTopic: PubsubTopic, contentTopic: ContentTopic): seq[PeerID] =
+  let filterCriterion : FilterCriterion = (pubsubTopic, contentTopic)
+
+  var foundPeers : seq[PeerID] = @[]
+  # only peers subscribed to criteria and with legit subscription is counted
+  s.subscriptions.withValue(filterCriterion, peers):
+    for peer in peers[]:
+      if s.isSubscribed(peer):
+        foundPeers.add(peer)
+
+  return foundPeers
+
+proc removePeer*(s: var FilterSubscriptions, peerId: PeerID) =
   ## Remove all subscriptions for a given peer
-  subscriptions.del(peerId)
+  s.peersSubscribed.del(peerId)
 
-proc removePeers*(subscriptions: var FilterSubscriptions, peerIds: seq[PeerID]) =
+proc removePeers*(s: var FilterSubscriptions, peerIds: seq[PeerID]) =
   ## Remove all subscriptions for a given list of peers
   for peerId in peerIds:
-    subscriptions.removePeer(peerId)
-
-proc containsAny*(criteria: FilterCriteria, otherCriteria: FilterCriteria): bool =
-  ## Check if a given pubsub topic is contained in a set of filter criteria
-  ## TODO: Better way to achieve this?
-  for criterion in otherCriteria:
-    if criterion in criteria:
-      return true
-  false
+    s.removePeer(peerId)
+
+proc cleanUp*(fs: var FilterSubscriptions) =
+  ## Remove all subscriptions for peers that have not been seen for a while
+  let now = Moment.now()
+  fs.peersSubscribed.keepItIf(now - val.lastSeen <= fs.subscriptionTimeout)
+
+  var filtersToRemove: seq[FilterCriterion] = @[]
+  for filterCriterion, subscribedPeers in fs.subscriptions.mpairs:
+    subscribedPeers.keepItIf(fs.isSubscribed(it)==true)
+
+  fs.subscriptions.keepItIf(val.len() > 0)
```

Comment:
Minor thing but I would put the clean-up in a timed loop instead of calling it when events happens.

That way you can "forget" about it.

Reply:
Well its there: https://github.com/waku-org/nwaku/blob/7e57b67933a21deac9d42db5571a6357cc0f54bb/waku/waku_filter_v2/protocol.nim#L183

There is a maintenance loop running every 1min...
But due to the nature of subscription limits we must remove from the store as expected the client will re-subscribe in a second call.

Reply:
Ah I thought we didn't have that NVM then, all good!